addons/web/static/src/core/file_upload/file_upload_service.js

```diff
@@ -0,0 +1,106 @@
+/** @odoo-module **/
+
+import { registry } from "@web/core/registry";
+import { sprintf } from "@web/core/utils/strings";
+
+const { reactive } = owl;
+
+export const fileUploadService = {
+    dependencies: ["notification"],
+    /**
+     * Overridden during tests to return a mocked XHR.
+     *
+     * @private
+     * @returns {XMLHttpRequest}
+     */
+    createXhr() {
+        return new window.XMLHttpRequest();
+    },
+
+    start(env, { notificationService }) {
+        const _uploads = reactive({});
+        let _nextId = 1;
+        /**
+         * @param {string}                          route
+         * @param {FileList}                        files
+         * @param {Object}                          [params]
+         * @param {function(formData): void}        [params.buildFormData]
+         * @param {Boolean}                         [params.displayErrorNotification]
+         * @param {EventBus}                        [params.bus]
+         * @returns {reactive}                      upload
+         * @returns {XMLHttpRequest}                upload.xhr
+         * @returns {FormData}                      upload.data
+         * @returns {Number}                        upload.progress
+         * @returns {Number}                        upload.loaded
+         * @returns {Number}                        upload.total
+         * @returns {String}                        upload.title
+         * @returns {String||undefined}             upload.type
+         */
+        const upload = async (route, files, params = {}) => {
+            const xhr = this.createXhr();
+            const bus = params.bus || env.bus;
```

Comment:
do you really need a bus here? please do not use the env.bus. your service should export a bus instead